staging/cse/windows/nvidiagpudriverfunc.ps1

```diff
@@ -0,0 +1,244 @@
+        Write-ConsoleLog $ErrorMsg
+        Set-ExitCode -ExitCode $global:WINDOWS_CSE_ERROR_SKU_INFO_NOT_FOUND -ErrorMessage $ErrorMsg
+    }
+      
+    $Setup = @{
+        RebootNeeded = ($vmSize -ne $null -and $Compute.vmSize -match "_NV")
+        Target       = "$RootDir\..\install.exe"
+    }
+
+    Get-DriverFile $GpuDriverURL $Setup.Target
+  
+    return $Setup
+}
+  
+function Get-VmData {
+    [OutputType([hashtable])]
+  
+    # Retry to overcome failure in retrieving VM metadata
+    $Loop = $true
+    $RetryCount = 0
+    $RetryCountMax = 10
+    do {
+        try {
+            $Compute = Invoke-RestMethod -Headers @{"Metadata" = "true" } -URI http://169.254.169.254/metadata/instance/compute?api-version=2017-08-01 -Method get
+            $Loop = $false
+        }
+        catch {
+            if ($RetryCount -gt $RetryCountMax) {
+                $Loop = $false
+                $ErrorMsg = "Failed to retrieve VM metadata after $RetryCountMax attempts. Exiting! More information on troubleshooting is available at https://aka.ms/NvidiaGpuDriverWindowsExtensionTroubleshoot"
+                Write-ConsoleLog $ErrorMsg
+                Set-ExitCode -ExitCode $global:WINDOWS_CSE_ERROR_SKU_INFO_NOT_FOUND -ErrorMessage $ErrorMsg
+            }
+            else {
+                $retryInSeconds = $RetryCount * 2 + 1
+                Write-ConsoleLog "Attempt $RetryCount of $RetryCountMax failed. Retrying in $retryInSeconds seconds."
+                Start-Sleep -Seconds ($retryInSeconds)
+                $RetryCount++
+            }
+        }
+    } while ($Loop)
+  
+    return $Compute
+}
+  
+function Get-DriverFile {
+    param(
+        [Parameter(Mandatory = $True)]
+        [ValidateNotNullOrEmpty()]
+        [string] $source,
+    
+        [Parameter(Mandatory = $True)]
+        [ValidateNotNullOrEmpty()]
+        [string] $dest
+    )
+  
+    Write-ConsoleLog "Downloading from $source to $dest"
+  
+    # Retry to overcome failure in downloading
+    $Loop = $true
+    $RetryCount = 0
+    $RetryCountMax = 10
+    do {
+        try {
+            Get-UsingWebClient -Url $source -OutputPath $dest
+            $Loop = $false
+            Write-ConsoleLog "Downloaded file successfully."
+            break
+        }
+        catch {
+            if ($RetryCount -gt $RetryCountMax) {
+                $Loop = $false
+                $ErrorMsg = "Failed to download $source after $RetryCountMax attempts. Exiting! More information on troubleshooting is available at https://aka.ms/NvidiaGpuDriverWindowsExtensionTroubleshoot"
+                Write-ConsoleLog $ErrorMsg
+                Set-ExitCode -ExitCode $global:WINDOWS_CSE_ERROR_GPU_DRIVER_INSTALLATION_DOWNLOAD_FAILURE -ErrorMessage $ErrorMsg
+            }
+            else {
+                Start-Sleep -Seconds ($RetryCount * 2 + 1)
+                $RetryCount++
+            }
+        }
+    } while ($Loop)
+}
+function Get-UsingWebClient {
```

Comment:
We use DownloadFileOverHttp so we can avoid download when the file is cached in the vhd.